### PR TITLE
Fix logical race caused by asynchronous view connection cleanup

### PR DESCRIPTION
**Pull request by JensAyton**
View connections are disposed asynchronously on the view queue (typically the main queue). If an event is dispatched on the view queue after stopping the Mobius controller, we would hit an assertion. Unlike previous logic races, this could reasonably be worked around in client code, but it would introduce annoying coupling and synchronization work which we can handle generically using our existing dropping-events-on-the-floor technique.

@jeppes

### CODE REVIEW

File: MobiusCore/Test/MobiusControllerTests.swift [hunk 1]
```diff
@@ -142,6 +142,17 @@ class MobiusControllerTests: QuickSpec {
                             controller.start()
                             expect(controller.running).to(beTrue())
                         }
+                        fit("should ignore events sent while view disposal is pending") {
```
Comment (rastersize) [inline MobiusCore/Test/MobiusControllerTests.swift]: This should probably be `it` and not `fit` for master 🙂

File: MobiusCore/Source/ConnectableConvenienceClasses/AsyncDispatchQueueConnectable.swift [hunk 2]
```diff
@@ -44,17 +54,33 @@ final class AsyncDispatchQueueConnectable<InputType, OutputType>: Connectable {
     }
 
     func connect(_ consumer: @escaping (OutputType) -> Void) -> Connection<InputType> {
-        let connection = underlyingConnectable.connect(consumer)
+        let connection = underlyingConnectable.connect { [disposalStatus] value in
+            // Don’t forward if we’re currently waiting to dispose the connection.
+            //
+            // NOTE: the underlying consumer must be called inside the criticial region accessing disposalStatus, or we
```
Comment (pettermahlen) [inline MobiusCore/Source/ConnectableConvenienceClasses/AsyncDispatchQueueConnectable.swift]:
```suggestion
            // NOTE: the underlying consumer must be called inside the critical region accessing disposalStatus, or we
```

File: MobiusCore/Source/ConnectableConvenienceClasses/AsyncDispatchQueueConnectable.swift [hunk 3]
```diff
@@ -44,17 +54,33 @@ final class AsyncDispatchQueueConnectable<InputType, OutputType>: Connectable {
...
 
         return Connection(
             acceptClosure: { [acceptQueue] input in
                 acceptQueue.async {
                     connection.accept(input)
                 }
             },
-            disposeClosure: { [acceptQueue] in
+            disposeClosure: { [acceptQueue, disposalStatus] in
+                disposalStatus.value = .pendingDispose
```
Comment (pettermahlen) [inline MobiusCore/Source/ConnectableConvenienceClasses/AsyncDispatchQueueConnectable.swift]: What happens if there are multiple calls to `dispose`?

File: MobiusCore/Test/MobiusControllerTests.swift (hunk 1 above)
Comment (JensAyton) [reply MobiusCore/Test/MobiusControllerTests.swift]: I was just thinking yesterday that we should lint for this…

File: MobiusCore/Source/ConnectableConvenienceClasses/AsyncDispatchQueueConnectable.swift (hunk 3 above)
Comment (JensAyton) [reply MobiusCore/Source/ConnectableConvenienceClasses/AsyncDispatchQueueConnectable.swift]: Updated to assert on that… and scope the `disposalStatus` to the individual connection, not the connectable. Oops!

File: MobiusCore/Test/MobiusControllerTests.swift (hunk 1 above)
Comment (rastersize) [reply MobiusCore/Test/MobiusControllerTests.swift]: @JensAyton We lint for it in S4A using [Danger](https://danger.systems) and the following snippet:

```ruby
# Make sure we aren't focusing a specific test and ignoring all others. We don't want to let
# testing shortcuts get into master.
(git.modified_files + git.added_files).each do |file|
  next unless File.file?(file)
  if file.downcase.include?('/tests/') && file.end_with?('.swift')
    contents = File.read(file)
    fail("`xit` or `fit` left in tests (#{file})") if contents =~ /^\w*[xf]it/
    fail("`fdescribe` left in tests (#{file})") if contents =~ /^\w*fdescribe/
    fail("`fitBehavesLike` left in tests (#{file})") if contents =~ /^\w*fitBehavesLike/
  end
end
```

**Pull request merged**